### DISCUSSION TITLE
RP2040 need to drive a 4V LED

### DISCUSSION THREAD

**Discussion by SezaroSteve** [ZuluSCSI/ZuluSCSI-firmware#241] [Q&A] [answered]
Hello,

The HDD LED in my Amiga 3000D operates on 4V. I did not know this at the time, so I hooked up my RP2040 (soldered two pins on the LED header) just like I have done in my other Amiga's and.... nothing. No HDD Activity LED.
As a reference, I measured the power LED and it works on 4V. As the HDD LED is of the same type I became suspicious that the Zulu's 3V is to low to make the LED work at all.

I attached a "lab power-supply" to the LED. The PSU can supply 0 - 15V up to 3 AMP very granularly and with this, I confirm my own suspicion. These original "Commodore" LED's need a starter Voltage of 3.65 V.
At 3.65 V, the LED starts to shine very faintly. At 4.0 V the HDD LED reaches normal brightness.

So the HDD LED is good (pfew..) but I need to feed it 4.0 V to work and the ZuluSCSI only provides 3.0 V.
I know how to reduce a voltage by placing a resistor inline on the + wire. But I have no idea how to **safely** increase the voltage from 3 to 4 V without damaging the Zulu.

With my limited knowledge of electronics, I found these small Adjustable Step Up DC power converters that can do this, but are they safe, as most of these units, take the + and - from the source, step it up and output a higher Voltage. They only have 4 pins: 2x "in" and 2x "out". But I guess this puts a strain on the source (higher Amps? I dunno).

Examples:
https://www.techtonics.in/dc-dc-3v-35v-to-4v-40v-adjustable-step-up-power-module-high-power-boost-converter

https://www.ebay.ch/itm/174959053725?hash=item28bc5f4b9d:g:gqUAAOSwhgpia46a&amdata=enc%3AAQAIAAAAwMbMfc9bc4VVBqrtSA5pkOTTya5hcfk4RBfmtPDpxqBeJ7tSEjU2F8s3pov3s11QFcenpY0UIjIPHTYipOGjWG819OfoXetr6gk%2BbfLoYnR5ukBS9CcwGGXsZz8P6mGHFbqzYKIZcEwo2pVbtvtN98xZucVU%2BHsxz621qDPPqJwAzQm7nh772d2ckP6gZfm6UvN0ScAEgwOKfNXVBjn%2FUk0ElHRaGoH4bjglTHQRE6JQpUBMjcsitM6O7tBY9yn%2FLA%3D%3D%7Ctkp%3ABk9SR5TBpfmVYg

**Comment by PetteriAimonen** [comment]
If you have +5V available in the system, easiest way is to use a transistor to control the LED.
The transistor can be any small NPN transistor.
Common types such as 2N2222, 2N3904 or BC547 are suitable.

[This tutorial](http://lednique.com/driving-leds-on-higher-voltage/) explains more of the details.
The basic circuit is:

![image](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/922265/bb6ac48a-1cf4-4b88-9509-6bfdae54076c)

R1 is from the ZuluSCSI LED output to the transistor base and can be any value between 1 kilo-ohm to 10 kilo-ohms. The Amiga LED probably has the LED resistor R2 already integrated.

In your case VSS would be the +5V supply from the computer, while ZuluSCSI would supply 3V to the transistor (0.7 V is enough to trigger NPN transistor).

**Reply by SezaroSteve** [reply]
Thanks for the reply.

The Amiga 3000D motherboard delivers 5V to the "LED Board" (contains the power and HDD LED's) and a resistor on the motherboard, which sits just before the pin-header, reduces the voltage to 4.0V. Thus your assumption "The Amiga LED probably has the LED resistor R2 already integrated" is basically true.
In my case though, I isolated the HDD LED from this construct. It is physically still on the LED Board but does not share anything with the circuitry on it anymore (traces cut) so it's now "stand alone" and I soldered wires to the HDD LED's Cathode and Anode and route it towards the Zulu so to speak.

Let's see if my non-electronics brain understands the concept:
- The Transistor's "base" pin is directly connected to the Zulu LED-Header's 3.0 V "Plus" (called output in the schematic above)
- R1 is not needed as the voltage coming from the Zulu is already the desired 3.0 V ?
   (I don't understand why R1 would need to be between 1 kilo-ohm to 10 kilo-ohms)
- VSS is +5V which comes from the computer. I'll tap this off from the 5V Rail that's on a floppy connector for example.
- R2 is a standard inline resistor that brings the voltage down to 4.0 V
- R2 continues on towards Anode + of the LED (the voltage that arrives here is 4.0 V )
- The Cathode of the LED (the - pin) continues on towards the "collector" pin of the transistor
- The transistor's "emitter" pin is connected to the ground-wire of the same 5V Rail (floppy connector)
- The Zulu's LED-Header's ground-wire is connected to the same ground-wire

If the workflow above is correct, how much Ohm must R2 be to drop +5V to +4V

I remember from School (long ago) that resistance (Ohm) is temperature dependent. The LED must never ever get fed more than 4.0 V so would it be wise to insert a voltage-regulator (max. 4.0V) between R2 and the LED's + (anode) so that no matter what happens, it never receives more than 4.0 V.
In my classic car, I have such a voltage regulator in the instrument-panel and it ensures that no matter what voltage comes from the car (can range from 11V to 13.6V) , the voltage that enters the instrument-panel with the gauges etc, never ever exceeds 10V so they don't get damaged.
In my case now, it would have a fixed output of 4.0V. Now that I think of it, I might as well replace R2 with such a voltage-regulator correct? (I would achieve dropping VSS to 4.0V and also be protected against over-voltages)

**Reply by PetteriAimonen** [reply]
R1 is needed and should be between 1-10kohm.
Otherwise the transistor base gets too much current.

A good guess for LED current is about 5mA, which would mean a 200 ohm resistor to drop 1 volt.
Series resistors are a well established way to control LED current, I wouldn't bother with voltage regulators.

**Reply by SezaroSteve** [reply]
These LED's are large square LED's from the 90's. I have no idea about their specs. Maybe it is possible to deduct the specs by means of this official schematic of the Amiga 3000D LED Board ?

To avoid confusion:  as you can see, the A3000D LED's normally work the other way around: The + on both LED's is permanently connected (red wire) and their (black) ground wires are connected only at times of activity (completing the circuit).
The Power LED (called  _LED ) and the HDD LED Circuit is completed by those "7407" things (I have no idea what they are)
The power LED is always on but it has two modes: depending on the state of the audio low-pass filter, it is either dimmed (multi-meter then reads 3.82 V) or at normal brightness (multi-meter then reads 4.04 V).

The "SCSI_LED" (the HDD LED that i'm talking about) is now completely disconnected from all of this circuitry. It is physically still on the board but the traces are cut and it's now totally stand-alone with a wire soldered to the LED's + and another wire soldered to the LED -, then both wires are routed towards where the Zulu sits in the case etc. 

![3000D LED-Board Schematic](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/4e434adc-5874-4d25-b3e9-65ead88ac54f)

The "Red Wire" comes from Motherboard header Pin 1 and feeds both Power LED and HDD LED simultaneously.
Pin 3 on the header goes to the ground pin of the Power LED.
Pin 2 on the header goes to the ground pin of the HDD LED this is now removed. Only Pins 1 and 3 are co

Maybe this schematic provides enough information to calculate the LED current of these LED's ?

Could it work like this: the Power LED Voltage of +5V is normally reduced to +4V by resistor R301 (200 Ohm) and then, when the Audio bypass filter state is changed, that 7407 thingy kicks in, completing the circuit there too, creating a total resistance of 268 Ohm as R307 now gets involved, which drops the voltage a bit more which dims the brightness of the Power LED a bit ?
(I don't know if these two resistors are in serial or parallel when R307 gets activated)

**Reply by PetteriAimonen** [reply]
With that driver, you could just connect ZuluSCSI LED output to the SCSI_LED input and reuse the driver on board.

But yeah, 200 ohm will work fine as the LED resistor.

**Reply by SezaroSteve** [reply]
"With that driver, you could just connect ZuluSCSI LED output to the SCSI_LED input and reuse the driver on board."

I'm sorry but I don't know what you mean. The "driver on board"?

**Reply by PetteriAimonen** [reply]
The driver shown in the schematic.

**Reply by SezaroSteve** [reply]
You mean the +5V on Pin 1 of the motherboard header? (otherwise I don't know what you mean with "driver", sorry for being such a noob)

**Reply by PetteriAimonen** [reply]
The 7407 driver IC which is used to control the LED originally can take input from ZuluSCSI.

**Reply by SezaroSteve** [reply]
> The 7407 driver IC which is used to control the LED originally can take input from ZuluSCSI.

I understand now. But i'm not going to touch the motherboard. I'm leaving everything there alone. The Motherboard was rebuilt/restored at great financial- and labor-expense and i'm not going to solder a bodge-wire onto some chip's leg. The entire solution must not touch the orginal Amiga at all.
I'll gladly steal +5V from a Molex or Floppy connecter to feed the base-pin of the transistor, that's fine. Everything else is purely Zulu -> LED.

I'll be building a test-rig for this first. I will not touch the original HDD LED before I am 100% sure that it will not get damaged.
The LED's don't exist anymore (they are 10mm wide in one-piece) and if damage that LED, i'm in big poopoo as replacements are impossible to get (or insanely expensive).
(there are several after-market LED "3000D LED-boards" with several flat LED's besides each other to mimic the 10mm wide original which look terrible imho).

I'm now looking for a LED that can take the place of the original 4V HDD-LED's (has a forward voltage of 3.6V) on the test-rig. I can't find LED's that are specced for 4V **and also** have a current of about 5mA.
They are all 3.3V or 5V and 10 or 20 mA nowadays.
In my test-rig, I could take a modern, round, 5V 20mA LED and use a 50 Ohm resistor to reduce 5V to 4V correct? (and assuming it has a forward-voltage of a little less than 4V)

(I sound more knowledgeable than I am. I learned about Ohm Law and what a transistor is, only today).

**Reply by PetteriAimonen** [reply]
Ok, this is probably more than what can be resolved remotely. I thought that schematic was for a "LED board" not motherboard.

Maybe you can seek out a friend or someone else local who has more electronics experience and can guide you through this?

**Reply by SezaroSteve** [reply]
That schematic IS for the A3000 LED Board. The 4707 Chip is on the motherboard. The LED-Board has no IC's on it.
![Screenshot 2023-05-01 at 21 10 29](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/a20820aa-b48e-423e-ba90-25a49e9e78ad)

This is what the A3000D LED-Board looks like. There are two generations (red and green PCB) but they are the same otherwise.

**Comment by SezaroSteve** [comment]
Thanks for putting up with me this long. I don't have anyone that can help me so i'll have to figure it out. The test-rig should allow me to blow up as many LED's as needed until I get it right.

**Comment by SezaroSteve** [comment]
One last question if I may:  where does the ground-wire from the Zulu goto? It's +3V will go to R1, but it's ground?

**Reply by PetteriAimonen** [reply, marked as answer]
All grounds should be connected together. They probably are already connected together through the power supply anyway.

**Reply by aperezbios** [reply]
@SezaroSteve the circular-looking plated through hole labeled 'C' (for cathode), is the right-most pin on the LED pin header, J304. The square-looking plated through-hole pin to the left, closest to the "LED" silk-screen, is the positive pin.

**Reply by SezaroSteve** [reply]
Hi yes I know. That's not what I was asking :-)
The + from the Zulu goes to R1. (normally this ends up at the + pin of the LED)
The - from the Zulu goes where?  (on the other side, not the Zulu Side). (normally this ends up at the - pin of the LED)
Petteri answered the question thnx.

**Comment by SezaroSteve** [comment]
Succes ! 

I ordered a 2N3904 NPN Resistor as per your advice and hooked everything up according to the schematic. And it works perfectly.

I used a modern-day 5V 20mA yellow 5mm round LED at first. When this worked, I lowered the 5V on my "lab powersupply" (Vss) and created traffic by running a disk benchmark. I then slooowly upped the Vss Voltage until the HDD LED on the LED-Board started to work.

As the exact, precise specs of the original yellow HDD LED are unknown, I was very careful and increased the "R2 resistor" bit by bit while also increasing the VSS voltage in turn, until the desired brightness was achieved.

I literally used the LED's brightness as the means to determine the correct resistor. The HDD LED lights up to almost the same brightness as the Power LED at normal brightness, which is the result I was after.

Long story short: I ended up with a 50k Ohm resistor connected to the base-pin of the transistor (R1 in the schematic).
Between the powersupply and the LED, I ended up using a 460 Ohm resistor on R2 with a Vss Voltage of 5.04 Volt as this is the voltage that the 5V Rail in my A3000D delivers.
The brightness is the result of the Ohm value of the resistor.

The brightness is a bit hard to compare as the HDD-LED flickers (I copied a large file between two (virtualized) SCSI HDD's to make the LED light up). But you get the idea.

Next thing to do is to package it nicely in a clean solution that can simply be plugged in between a Zulu's LED Header and the 3000D HDD LED. I'l report on that when I've built that too.

Time for Piccas: in one you see the HDD LED being lit up. In the second I drew up the final schematic.
![A3000D_ZuluSCSI_HDD-LED-Control_Prototype](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/70933f3b-c2bb-4303-a7d2-ebd1f5a43a47)
![A3000D_ZuluSCSI_HDD-LED-Control_Schematic](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/69816620-5510-4038-a914-f304229e5b7b)

**Reply by PetteriAimonen** [reply]
Nice to hear you got it solved and thanks for sharing the information!

**Comment by SezaroSteve** [comment]
And here is the finished product. I call it the "A3000 Desktop ZuluSCSI LED Adapter thingy".

I use a 3-Pin Case-fan style connector to connect/disconnect the adapter from the HDD LED ( I only use two of the 3 ping though).
It works perfectly :D

![A3000D-ZuluSCSI_LED-Adapter_1](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/6f263f31-c929-47bc-8848-b307f1aab3d9)

![A3000D-ZuluSCSI_LED-Adapter_2](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/123663970/b3696bdd-73a6-48b0-9616-ea4d0c9978bc)
